### ISSUE TITLE
Implementation of Timestamp Renewal and Hash-Tree Renewal for RFC 4998 (Evidence Record Syntax) missing(?)

### ISSUE THREAD

**Issue by raubv0gel** [bcgit/bc-java#1173]
I cannot find any (high-level) implementation and/or tests for Timestamp Renewal and Hash-Tree Renewal as defined in [RFC 4998 5.2.](https://datatracker.ietf.org/doc/html/rfc4998#section-5.2). Did I miss that or is it actually missing?

**Comment by raubv0gel** [comment]
Verification according to [RFC 4998 5.3.](https://datatracker.ietf.org/doc/html/rfc4998#section-5.3) seems to be missing as well.

**Comment by raubv0gel** [comment]
For an example [org.bouncycastle.tsp.ers.ERSEvidenceRecord#validatePresent(byte[], java.util.Date)](https://github.com/bcgit/bc-java/blob/7f4a0a69c19c516c9c154776128e284e6816ecf8/pkix/src/main/java/org/bouncycastle/tsp/ers/ERSEvidenceRecord.java#L56) does not validate correctly if an `ArchiveTimestamp` has no `reducedHashtree` filed (it is allowed according to [RFC 4998 5.2.](https://datatracker.ietf.org/doc/html/rfc4998#section-5.2)). Furthermore it only validates the last time stamp and not the whole `ArchiveTimeStampChain` (see [RFC 4998 5.3](https://datatracker.ietf.org/doc/html/rfc4998#section-5.3)).

**Comment by raubv0gel** [comment]
@dghgit, pardon, but do you have any thoughts to this issue? (I’m just guessing that you are the right person to contact.)

**Comment by raubv0gel** [comment]
@dghgit, thanks for the advice! I’ll try it.

**Comment by raubv0gel** [comment]
Commit for time stamp renewal and hash tree renewal is done: [https://github.com/raubv0gel/bc-java/tree/feature-ers-renewal](https://github.com/raubv0gel/bc-java/tree/feature-ers-renewal)

**Comment by dghgit** [comment]
I had a look. The TimeStamper interface won't work for this. Changing the API is problematic in itself, but the real issue is getting a timestamp in practice is highly asynchronous and may involve batching, errors, and retries.

**Comment by dghgit** [comment]
I've added TimeStamp renewal. There's an example of it in:

https://github.com/bcgit/bc-java/blob/0554530a86e3ca46149c810f4ab315cd1fac15f0/pkix/src/test/java/org/bouncycastle/tsp/test/ERSTest.java#L326

Looking at hash renewal now.

**Comment by raubv0gel** [comment]
@dghgit, thank you very much! I’ll have a look very soon …

**Comment by raubv0gel** [comment]
I looked at `org.bouncycastle.tsp.ers.ERSEvidenceRecord#buildTspRenewalGenerator`. Is it really a good idea to `ERSDataGroup timestampGroup = new ERSDataGroup(prevTimes);`? [5.2. Generation](https://datatracker.ietf.org/doc/html/rfc4998#section-5.2) says

> In the case of Timestamp Renewal, the content of the timeStamp field
   of the old Archive Timestamp has to be hashed and timestamped by a
   new Archive Timestamp.  The new Archive Timestamp MAY not contain a
   reducedHashtree field, if the timestamp only simply covers the
   previous timestamp.  **However, generally one can collect a number of
   old Archive Timestamps and build the _new hash tree_ with the hash
   values of the content of their timeStamp fields.**

For me it seems to be too complex to build a new hash tree for each time stamp renewal. If we create a new hash tree, we had to create _multiple_ `ArchiveTimeStamp`s, hence _multiple_ new `EvidenceRecord`s, right? And keep in mind, that in general we have _multiple_ `EvidenceRecord`s with the same time stamp that will need a renewed `ArchiveTimeStamp`. Adding a new simple `ArchiveTimeStamp` that covers the previous seems to be more easy (one can add this renewed `ArchiveTimeStamp` to _multiple_ `EvidenceRecord`s).

However, if one wants to create a _new hash tree_ covering _all_ previous `ArchiveTimeStamp`s, one should create a proper hash tree. `atsGen.addData(timestampGroup);` adds a _single_ `object data group` only, right? The hash tree would only have a single `data object  group` hash value leave (the only reduced hash tree would be `rht = SEQ( SEQ(ts1, ts2, ts3, …, tsn)  )` where `tsi` are the previous time stamps).

**Comment by raubv0gel** [comment]
Maybe helpful

![Hash-Tree](https://user-images.githubusercontent.com/9824146/175811529-16485b04-1a59-40a3-bc11-4fedb1f1f905.svg)

**Comment by raubv0gel** [comment]
This may help:

![Hash-Tree](https://user-images.githubusercontent.com/9824146/175811569-caa51332-37a3-4b55-b38e-a326d5302e07.png)

**Comment by raubv0gel** [comment]
### TimeStamper Interface

I added this interface because in both functions

``` java
public TimeStampRequest generateTimeStampRequest(TimeStampRequestGenerator tspReqGenerator, BigInteger nonce){};
```

and

``` java
public ERSArchiveTimeStamp generateArchiveTimeStamp(TimeStampResponse tspResponse){};
```

we see

``` java
PartialHashtree[] reducedHashTree = getPartialHashtrees();
// …
byte[] rootHash = rootNodeCalculator.computeRootHash(digCalc, reducedHashTree);
```

Because we need to call `generateTimeStampRequest()` and `generateArchiveTimeStamp()` one after another, we actually call `getPartialHashtrees()` and `computeRootHash()` twice. Both functions are computational complex (heavy hash value calculations). Hence, we unnecessarily doubled the time of calculation!

> … the real issue is getting a timestamp in practice is highly asynchronous and may involve batching, errors, and retries.

That’s right and thus my TimeStamper Interface is not enough. I guess, we need a class that stores the internal state of `generateTimeStampRequest()` (the result of `generateTimeStampRequest()` and/or the result of `computeRootHash()`). We than can use this state in `generateArchiveTimeStamp()` (no need to recalculate hash values).

Btw. this analogously holds for the time stamp renewal and hash tree renewal functions.

**Comment by dghgit** [comment]
Yes, I think finding some way to deal with the internal state is the trick. I've managed to cut a few corners with the computation as well.

With the time stamp renewal, I think they do it that way to ensure the provenance of the replacement timestamp - it's got a record of the entire history. As you guessed though using a data group then becomes the only way of doing it as otherwise you'd have multiple ArchiveTimeStamps due to the multiple reduced hash trees. At least in this case they do say there should just be one ArchiveTimeStamp so it's "obvious" what needs to be done.

You're correct as well that the timestamp response can be reused if the other evidence records are still in sync - I've included re-use in the test cases and it works fine.

Hash renewal seems to be a complete break though - as the different evidence records have different data in them it results in a break occurring for what used to be related data.

**Comment by raubv0gel** [comment]
Thank you! I will have a look …

> Hash renewal seems to be a complete break though - as the different evidence records have different data in them it results in a break occurring for what used to be related data.

I’m currently reviewing this. My guessing is the same, hash tree renewal seems to be broken by (RFC) design. But, I have to think more about it.

I will report back.

**Comment by dghgit** [comment]
I guess the problem with hash renewal is you really do have to have access to the original data - in some ways thinking about this does make me sympathetic to the decision to split the Merkle tree across different evidence records. It does break the problem of getting the original data up and there may be many reasons why dealing with or processing all the original data might be a non-technical issue.

**Comment by raubv0gel** [comment]
### Hash Tree Renewal

The current implementation `org.bouncycastle.tsp.ers.ERSEvidenceRecord#renewHash` does _not_ correspond to [5.2. Generation](https://datatracker.ietf.org/doc/html/rfc4998#section-5.2) actually. To generate a new hash tree we have to rehash _all_ original `data object`s `d_i` and concatenate the DER encoded `ArchiveTimeStampSequence` of `ER(d_i)`. That is, we need _all_ original `data object`s and _all_ corresponding `EvidenceRecord`s. The current implementation rehashes only a _single_ `data object` (`ERSData data`) …

> My guessing is the same, hash tree renewal seems to be broken by (RFC) design.

Now, I think, it’s _not_ broken – implementation should be possible.

**Comment by dghgit** [comment]
I wondered that initially, it's quite explicit though:

Section 5.2, Point 2: Select data objects d(i) referred to by initial Archive Timestamp (objects that are still present and not deleted).  Generate hash values h(i) = H((d(i)).

This means the objects in the first partial hash tree.

The other thing I realised it doesn't make sense to do so either - because the hash of the the ArchiveTimeStampSequence is also included, even if all the objects were present, you still have to time stamp each evidence record from the original set individually as the root of the Merkle tree will never be the same so adding all the original objects as well would just create a lot of work for nothing (you can't tell if they're there as they are not in the primary partial hash tree and only the primary time stamp will tell you that two ERs are related).

Yes, full validation needs more work.

**Comment by raubv0gel** [comment]
@dghgit, does it help if I write a new, more complex, unit test? Let’s say, like in the tree image (with data d1 to d8) above? I would try to choose data with easy to understand hash values (in the sense of sorting).

Furthermore, I can possibly do a code review if the implementation is “complete”. Would that be desired?

**Comment by dghgit** [comment]
@raubv0gel absolutely "yes" to both. I've also put up a beta release in https://www.bouncycastle.org/betas if it's any help.

**Comment by dghgit** [comment]
Okay, I think one is settled. Still waiting on the test data #1164 but I will close this one off now.

**Comment by raubv0gel** [comment]
@dghgit, I will add a new issue to provide some test data: #1216
